src/database-controller/src/write-merger/handler.js

```diff
@@ -0,0 +1,287 @@
+        ) {
+          // if framework request is equal, mark requestSynced = true
+          logger.info(`The request of framework ${frameworkName} is synced.`);
+          internalUpdate.requestSynced = true;
+        } else {
+          // if framework request is not equal,
+          // should use framework request in db as ground truth
+          internalUpdate.requestSynced = false;
+        }
+        // use request in database
+        snapshot.overrideRequest(oldSnapshot);
+        if (req.params.eventType === 'DELETED') {
+          if (snapshot.getState() === 'Completed') {
+            // if event is DELETED and the state is Completed, mark apiServerDeleted = true
+            internalUpdate.apiServerDeleted = true;
+          } else {
+            // Event is DELETED and the state is not Completed.
+            // This case could occur when someone deletes the framework in API server directly.
+            // In such case, we mark requestSynced=false, and reset the snapshot using the framework request.
+            snapshot = new Snapshot(snapshot.getRequest(false));
+            internalUpdate.requestSynced = false;
+            internalUpdate.apiServerDeleted = false;
+          }
+        }
+        await databaseModel.Framework.update(
+          _.assign(snapshot.getStatusUpdate(), internalUpdate),
+          { where: { name: frameworkName } },
+        );
+      }
+    });
+    res.status(200).json({ message: 'ok' });
+  } catch (err) {
+    return next(err);
+  }
+}
+
+async function onCreateFrameworkRequest(snapshot, submissionTime, addOns) {
+  // create new record in db
+  // including all add-ons and submissionTime
+  // set requestGeneration = 1
+  snapshot.setRequestGeneration(1);
+  const record = _.assign(
+    { requestSynced: false, apiServerDeleted: false },
+    snapshot.getAllUpdate(),
+    addOns.getUpdate(),
+  );
+  record.submissionTime = new Date(submissionTime);
+  await databaseModel.Framework.create(record);
+  // Poller has an interval to synchronize the framework request to API server.
+  // The interval will cause a delay on every job.
+  // So we introduce a short-cut here to synchronize the framework request.
+  // It is an async function call and doesn't affect the return of this function.
+  // Any error of it will be logged and ignored.
+  silentSynchronizeRequest(snapshot, addOns);
+}
+
+async function onModifyFrameworkRequest(oldSnapshot, snapshot, addOns) {
+  // compare framework request (omit requestGeneration)
+  if (_.isEqual(snapshot.getRequest(true), oldSnapshot.getRequest(true))) {
+    // request is equal, no-op
+  } else {
+    // request is different
+    // update request in db, mark requestSynced=false
+    snapshot.setRequestGeneration(oldSnapshot.getRequestGeneration() + 1);
+    await databaseModel.Framework.update(
+      _.assign({}, snapshot.getRequestUpdate(), {
+        requestSynced: false,
+      }),
+      { where: { name: snapshot.getName() } },
+    );
```

Comment:
If databaseModel.Framework.update failed, will silentSynchronizeRequest still be called?

Make sure silentSynchronizeRequest is called only when DB success

Reply:
This is guaranteed. `await` will throw any error that happened.